prog/kube-peers/main.go

```diff
@@ -52,20 +40,109 @@ func getKubePeers() ([]string, error) {
 
 		// Fallback for cases where a Node has an ExternalIP but no InternalIP
 		if internalIP != "" {
-			addresses = append(addresses, internalIP)
+			addresses = append(addresses, nodeInfo{name: peer.Name, addr: internalIP})
 		} else if externalIP != "" {
-			addresses = append(addresses, externalIP)
+			addresses = append(addresses, nodeInfo{name: peer.Name, addr: externalIP})
 		}
 	}
 	return addresses, nil
 }
 
+const (
+	configMapName      = "weave-net"
+	configMapNamespace = "kube-system"
+)
+
+// update the list of all peers that have gone through this code path
+func addMyselfToPeerList(c *kubernetes.Clientset, peerName, name string) (*peerList, error) {
+	cml := newConfigMapAnnotations(configMapNamespace, configMapName, c)
+	var list *peerList
+	err := cml.LoopUpdate(func() error {
+		var err error
+		list, err = cml.GetPeerList()
+		if err != nil {
+			return err
+		}
+		log.Println("Fetched existing peer list", list)
+		if !list.contains(peerName) {
+			list.add(peerName, name)
+			log.Println("Storing new peer list", list)
+			err = cml.UpdatePeerList(*list)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return list, err
+}
+
+// For each of those peers that is no longer listed as a node by
+// Kubernetes, remove it from Weave IPAM
+func reclaimRemovedPeers(apl *peerList, nodes []nodeInfo) error {
+	// TODO
+	// Outline of function:
+	// 1. Compare peers stored in the peerList against all peers reported by k8s now.
```

Comment:
For info:
Our hacky rmpeers solution was hitting this quite a bit. So this is our current fix:
```
      initContainers:
      - name: reset-weavedb
        image: hub.docker.tech.lastmile.com/ocadotechnology/weave-net-reset-weavedb:0acd87fd3f7bde7d4cfe8b5009e0fafa2d0e9c18@sha256:05722937400c0e0449a80e9af8285323ac9ca7036c7f009e6c3c1e46d4bae855
        env:
        - name: HOSTNAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
        volumeMounts:
        - name: weavedb
          mountPath: /weavedb
```

The source of which is:
```
#!/bin/sh

set -eux

CURRENT_NODE_UUID=$(kubectl get node -o custom-columns=ID:.metadata.uid --no-headers ${HOSTNAME})

if [ ! -f /weavedb/NODE_UUID ]; then
  # Does not exist, so create it with the Node's current UUID
  echo -n ${CURRENT_NODE_UUID} > /weavedb/NODE_UUID
  exit 0;
fi;

if [ "${CURRENT_NODE_UUID}" == "$(cat /weavedb/NODE_UUID)" ]; then
  # It's still the same, so we're not likely to be out of sync
  exit 0;
fi;

# If we're here, it didn't match, so reset the persistent data on the assumption
# that we've been rmpeer'd
rm -f /weavedb/weavedata.db
rm -f /weavedb/weave-netdata.db

# And also set the NODE_UUID
echo -n ${CURRENT_NODE_UUID} > /weavedb/NODE_UUID

exit 0
```

We're basing this of the assumption that if our Node object has gone away and come back with a new UUID, our IPs may have been hoovered up by our rmpeers script, so we'll start over.